libtransmission/crypto.cc

```diff
@@ -4,144 +4,168 @@
 // License text can be found in the licenses/ folder.
 
 #include <cstring> /* memcpy(), memmove(), memset() */
+#include <string_view>
 
 #include <arc4.h>
 
+#include <math/wide_integer/uintwide_t.h>
+
+#include <arpa/inet.h> // htonl
+
 #include "transmission.h"
 #include "crypto.h"
 #include "crypto-utils.h"
 #include "tr-assert.h"
-#include "utils.h"
 
-///
+using namespace std::literals;
 
-tr_crypto::tr_crypto(tr_sha1_digest_t const* torrent_hash, bool is_incoming)
-    : is_incoming_{ is_incoming }
+namespace wi
 {
-    if (torrent_hash != nullptr)
-    {
-        this->torrent_hash_ = *torrent_hash;
-    }
-}
+using key_t = math::wide_integer::uintwide_t<
+    tr_message_stream_encryption::KeySize * std::numeric_limits<unsigned char>::digits>;
 
-tr_crypto::~tr_crypto()
+using private_key_t = math::wide_integer::uintwide_t<
+    tr_message_stream_encryption::PrivateKeySize * std::numeric_limits<unsigned char>::digits>;
+
+// source: https://stackoverflow.com/a/1001373/6568470
+// nb: when we bump to std=C++20, use `std::endian`
+bool is_big_endian()
 {
-    tr_dh_secret_free(my_secret_);
-    tr_dh_free(dh_);
-    tr_free(enc_key_);
-    tr_free(dec_key_);
-}
+    union
+    {
+        uint32_t i;
+        char c[4];
+    } bint = { 0x01020304 };
 
-///
+    return bint.c[0] == 1;
+}
 
-namespace
+template<typename Integral>
```

Comment:
```suggestion
template<typename UIntWide>
```

both here and in `export_bits()` to make it explicit that this function requires a uintwide_t

Reply:
fixed in 4a1ac4ada18e4e6471631798b6afbfd8183e55a0